tools/git-change-exec/main.go

```diff
@@ -0,0 +1,307 @@
+	gitPath        string
+	g              *git.Repository
+	visitedPaths   map[string]struct{}
+}
+
+func debugLog(fmt string, args ...any) {
+	if debug {
+		log.Printf(fmt, args...)
+	}
+}
+
+func newGitChangeExec() gitChangeExec {
+	return gitChangeExec{
+		actionDos:      map[action]struct{}{},
+		actionsToCheck: []action{},
+		visitedPaths:   map[string]struct{}{},
+	}
+}
+
+func main() {
+
+	actionCategories := []string{}
+
+	for c := range actions {
+		actionCategories = append(actionCategories, c)
+	}
+
+	rootCmd := cobra.Command{
+		Args: cobra.MinimumNArgs(1),
+		Use:  strings.Join(actionCategories, "|"),
+		Run: func(cmd *cobra.Command, args []string) {
+			gce := newGitChangeExec()
+
+			for _, cat := range args {
+				if len(actions[cat]) == 0 {
+					log.Fatalf("could not find actions for %s", cat)
+				}
+
+				gce.actionsToCheck = append(gce.actionsToCheck, actions[cat]...)
+			}
+
+			currentPath, err := os.Getwd()
+			if err != nil {
+				log.Fatalf("getting current working directory: %v", err)
+			}
+
+			defer func() {
+				err := os.Chdir(currentPath)
+				if err != nil {
+					log.Printf("could not change back to previous dir %s: %v", currentPath, err)
+				}
+			}()
+
+			gce.g, err = git.PlainOpenWithOptions("./", &git.PlainOpenOptions{DetectDotGit: true})
+			if err != nil {
+				log.Fatalf("open git path %s failed: %v", gce.gitPath, err)
+			}
+
+			wt, err := gce.g.Worktree()
+			if err != nil {
+				log.Fatalf("could not determine worktree: %v", err)
+			}
+			gce.gitPath = wt.Filesystem.Root()
+
+			err = os.Chdir(gce.gitPath)
+			if err != nil {
+				log.Fatalf("could not change to %s: %v", gce.gitPath, err)
+			}
+
+			gce.fetchOrigin()
+
+			gce.collectActionsGitTree()
+			gce.collectDirtyGitTree()
+
+			gce.runActionDos()
+		},
+	}
+
+	rootCmd.PersistentFlags().BoolVarP(&dryRun, "dry-run", "d", false, "")
+	err := rootCmd.Execute()
+	if err != nil {
+		log.Fatalf("corba failed with: %v", err)
+	}
+
+}
+
+func (gce *gitChangeExec) fetchOrigin() {
+	err := gce.g.Fetch(&git.FetchOptions{
+		RemoteName: "origin",
+		Tags:       git.AllTags,
+	})
+	if err != nil {
+		debugLog("fetching from origin failed: %v", err)
+	}
+}
+
+func (gce *gitChangeExec) collectActionsGitTree() {
+
+	logIter, err := gce.g.Log(&git.LogOptions{})
+	if err != nil {
+		log.Fatalf("getting log failed: %v", err)
+	}
+
+	branchHead, err := logIter.Next()
+	if err != nil {
+		log.Fatalf("getting log.Next failed: %v", err)
+	}
+
+	commonBase := gce.findCommonBase(branchHead)
+
+	logIter, err = gce.g.Log(&git.LogOptions{})
+	if err != nil {
+		log.Fatalf("getting log for iteration failed: %v", err)
+	}
+
+	err = logIter.ForEach(func(c *object.Commit) error {
+		for _, cb := range commonBase {
+			if c.Hash == cb.Hash {
+				return storer.ErrStop
+			}
+		}
+
+		commitStats, err := c.Stats()
+		if err != nil {
+			log.Fatalf("getting commit stats failed: %v", err)
+		}
+
+		for _, st := range commitStats {
+			gce.addActionByPath(st.Name)
```

Comment:
I don't think it makes sense to add actions for files that don't exist in the current branch's HEAD.
Also, as far as I understand it, it will run for all the commits between the HEAD and the common base**S**, even the oldest, the most irrelevant `-stable`... It will include tens/hundreds of irrelevant commits, hence - files. Or did I get the algorithm wrong?

Reply:
It will stop if it finds a hash that is in at least one of the common base**S**